magda-web-client/src/Components/Dataset/Add/Pages/People/CustodianDropdown.tsx

```diff
@@ -0,0 +1,102 @@
+import React, { useState } from "react";
+import { useAsync } from "react-async-hook";
+import Select from "react-select";
+import { config } from "config";
+import find from "lodash/find";
+
+import ReactSelectStyles from "Components/Common/react-select/ReactSelectStyles";
+
+import {
+    listOrgUnitsAtLevel,
+    OrgUnitWithRelationship
+} from "api-clients/OrgUnitApis";
+
+type Props = {
+    orgUnitId?: string;
+    teamOrgUnitId?: string;
+    onChange: (orgUnitId: string) => void;
+};
+
+export default function CustodianDropdown({
+    orgUnitId,
+    teamOrgUnitId,
+    onChange: onChangeCallback
+}: Props) {
+    // --- use hasUserSelected to tell whether user has selected a value or not
+    const [hasUserSelected, setHasUserSelected] = useState(false);
+    const { loading, error, result, execute } = useAsync(listOrgUnitsAtLevel, [
+        config.custodianOrgLevel,
+        teamOrgUnitId
+    ]);
+
+    if (loading) {
+        return <span>Loading...</span>;
+    } else if (error || !result || result.length === 0) {
+        return (
+            <div className="au-body au-page-alerts au-page-alerts--error">
+                <span style={{ verticalAlign: "-2px" }}>
+                    Could not retrieve data custodian list, or there are no data
+                    custodians in the system.
+                </span>
+                <button
+                    className="au-btn au-btn--tertiary"
+                    onClick={() =>
+                        execute(config.custodianOrgLevel, teamOrgUnitId)
+                    }
+                >
+                    Try Again
+                </button>
+            </div>
+        );
+    } else {
+        const selectedValue = orgUnitId && find(result, option => option.id === orgUnitId);
+
+        if ((!selectedValue || !hasUserSelected) && teamOrgUnitId) {
+            const relatedOrgUnit = find(
+                result,
+                option =>
+                    option.relationship && option.relationship !== "unrelated"
+            );
+            if (
+                relatedOrgUnit &&
+                (relatedOrgUnit as OrgUnitWithRelationship).id !== orgUnitId
+            ) {
+                // --- we do need to send out onChangeCallback to update upper level state to make sure it's consistent with the screen value
+                // --- only update when we find a different value here otherwise we will trigger a infinite loop
+                onChangeCallback(
```

Comment:
Is this because we only know what the default should be after we've called the API to find out what custodian org unit the user has a relationship with? Could it be done with `useEffect(..., [results])` ?

Reply:
1> It's for the following scenario:
If we set a default value on screen and user is ok with the value, he probably won't touch the dropdown before click `next page` ---- for this case, the default value never saved to metaDataPage state.
2> What's the reason that we want to use a useEffect hook? 
You can tell this logic is under a conditional block. Thus, probably not good use case for hooks?

Reply:
The thing is it's equivalent to using `setState` in a render, which should [never](https://github.com/facebook/react/issues/5591#issuecomment-161678219) happen 😬 .

I was thinking the `useEffect` would sit at the top level of the render function and watch for changes to results - I think it'd also remove the need for `hasUserSelected`. Something like

```javascript
useEffect(() => {
    if (!orgUnitId) {
        onChangeCallback(getDefaultCustodianId(results, teamOrgUnitId))
    }
}, [results])
```

Does that make any sense? The idea being that because it's in a hook it wouldn't run on render, so there's no mixing the purpose of the render fn, and no (less?) danger of causing an infinite loop. The other alternative might be to do it as a `then` on the promise?